testing/test_basics.py

```diff
@@ -249,12 +249,25 @@ class Arg:
 
 @pytest.mark.skipif("not hasattr(os, 'dup')")
 def test_stdouterrin_setnull(execmodel, capfd):
-    gateway_base.init_popen_io(execmodel)
-    os.write(1, b"hello")
-    os.read(0, 1)
-    out, err = capfd.readouterr()
-    assert not out
-    assert not err
+    # Backup and restore stdin state, and rely on capfd to handle
+    # this for stdout and stderr.
+    orig_stdin = sys.stdin
+    orig_stdin_fd = os.dup(0)
+    try:
+        # The returned Popen2IO instance can be garbage collected
+        # prematurely since we don't hold a reference here, but we
+        # tolerate this because it is intended to leave behind a
+        # sane state afterwards.
+        gateway_base.init_popen_io(execmodel)
+        os.write(1, b"hello")
+        os.read(0, 1)
+        out, err = capfd.readouterr()
+        assert not out
+        assert not err
+    finally:
+        sys.stdin = orig_stdin
+        os.dup2(orig_stdin_fd, 0)
+        os.close(orig_stdin_fd)
```

Comment:
When I run the full test suite in a local venv including eventlet and gevent, for some reason without this stdin restoration test_stdouterrin_setnull tends trigger the "Bad file descriptor" error just for gevent which runs last. I don't understand why that happens even with `closefd=False` in init_popen_io, but anyway this stdin restoration appears to prevent it.

Reply:
there are some occasional issues with coordination of gevent - the patching of stdio with fdopen of gevent is trouble in some cases

Reply:
in particular in the cases where it collides with pytest stdio capture